std/experimental/allocator/building_blocks/region.d

```diff
@@ -878,3 +939,388 @@ version(Posix) @system unittest
+    }
+}
+
+@system unittest
+{
+    import std.experimental.allocator.mallocator : Mallocator;
+
+    static void testAlloc(Allocator)(ref Allocator a, bool growDownwards)
+    {
+        import core.thread : ThreadGroup;
+        import std.algorithm.sorting : sort;
+        import core.internal.spinlock : SpinLock;
+
+        SpinLock lock = SpinLock(SpinLock.Contention.brief);
+        enum numThreads = 100;
+        void[][numThreads] buf;
+        size_t count = 0;
+
+        void fun()
+        {
+            void[] b = a.allocate(63);
+            assert(b.length == 63);
+
+            lock.lock();
+            buf[count] = b;
+            count++;
+            lock.unlock();
+        }
+
+        auto tg = new ThreadGroup;
+        foreach (i; 0 .. numThreads)
+        {
+            tg.create(&fun);
+        }
+        tg.joinAll();
+
+        sort!((a, b) => a.ptr < b.ptr)(buf[0 .. numThreads]);
+        foreach (i; 0 .. numThreads - 1)
+        {
+            assert(buf[i].ptr + a.goodAllocSize(buf[i].length) == buf[i + 1].ptr);
+        }
+
+        assert(!a.deallocate(buf[1]));
+
+        foreach (i; 0 .. numThreads)
+        {
+            if (!growDownwards)
+                assert(a.deallocate(buf[numThreads - 1 - i]));
+            else
+                assert(a.deallocate(buf[i]));
+        }
+    }
+
+    auto a1 = SharedRegion!(Mallocator, Mallocator.alignment,
+        Yes.growDownwards)(1024 * 64);
+
+    auto a2 = SharedRegion!(Mallocator, Mallocator.alignment,
+        No.growDownwards)(1024 * 64);
+
+    testAlloc(a1, true);
+    testAlloc(a2, false);
+}
+
+@system unittest
+{
+    import std.experimental.allocator.mallocator : Mallocator;
+
+    static void testAlloc(Allocator)(ref Allocator a, bool growDownwards)
+    {
+        import core.thread : ThreadGroup;
+        import std.algorithm.sorting : sort;
+        import core.internal.spinlock : SpinLock;
+
+        SpinLock lock = SpinLock(SpinLock.Contention.brief);
+        enum numThreads = 100;
+        void[][2 * numThreads] buf;
+        size_t count = 0;
+
+        void fun()
+        {
+            void[] b = a.allocate(63);
+            assert(b.length == 63);
+
+            lock.lock();
+            buf[count] = b;
+            count++;
+            lock.unlock();
+
+            b = a.alignedAllocate(63, 32);
+            assert(b.length == 63);
+            assert(cast(size_t) b.ptr % 32 == 0);
+
+            lock.lock();
+            buf[count] = b;
+            count++;
+            lock.unlock();
+        }
+
+        auto tg = new ThreadGroup;
+        foreach (i; 0 .. numThreads)
+        {
+            tg.create(&fun);
+        }
+        tg.joinAll();
+
+        sort!((a, b) => a.ptr < b.ptr)(buf[0 .. 2 * numThreads]);
+        foreach (i; 0 .. 2 * numThreads - 1)
+        {
+            assert(buf[i].ptr + buf[i].length <= buf[i + 1].ptr);
+        }
+
+        assert(!a.deallocate(buf[1]));
```

Comment:
Maybe we should consider adding an `alignedDeallocate` method that is able to solve the issue of deallocating aligned buffers?